dll/win32/shimgvw/loader.cpp

```diff
@@ -0,0 +1,264 @@
+{
+    DWORD Transferred;
+    if (!ReadFile(hFile, Buffer, Size, &Transferred, NULL))
+        return HResultFromWin32(GetLastError());
+    return Size == Transferred ? S_OK : HResultFromWin32(ERROR_HANDLE_EOF);
+}
+
+struct IMAGEINFO
+{
+    UINT w, h;
+    BYTE bpp;
+};
+
+class BitmapInfoHeader : public BITMAPINFOHEADER
+{
+public:
+    BitmapInfoHeader() {}
+    BitmapInfoHeader(const void* pbmiHeader) { Initialize(pbmiHeader); }
+
+    void Initialize(const void* pbmiHeader)
+    {
+        BITMAPINFOHEADER& bih = *(BITMAPINFOHEADER*)pbmiHeader;
+        if (bih.biSize >= sizeof(BITMAPINFOHEADER))
+        {
+            CopyMemory(this, &bih, min(bih.biSize, sizeof(*this)));
+        }
+        else
+        {
+            ZeroMemory(this, sizeof(*this));
+            BITMAPCOREHEADER& bch = *(BITMAPCOREHEADER*)pbmiHeader;
+            if (bih.biSize >= sizeof(BITMAPCOREHEADER))
+            {
+                biSize = bch.bcSize;
+                biWidth = bch.bcWidth;
+                biHeight = bch.bcHeight;
+                biPlanes = bch.bcPlanes;
+                biBitCount = bch.bcBitCount;
+                biCompression = BI_RGB;
+            }
+        }
+    }
+};
+
+static inline bool IsPngSignature(const void* buffer)
+{
+    const BYTE* p = (BYTE*)buffer;
+    return p[0] == 0x89 && p[1] == 'P' && p[2] == 'N' && p[3] == 'G' &&
+           p[4] == 0x0D && p[5] == 0x0A && p[6] == 0x1A && p[7] == 0x0A;
+}
+
+static inline bool IsPngSignature(const void* buffer, SIZE_T size)
+{
+    return size >= 8 && IsPngSignature(buffer);
+}
+
+static BYTE GetPngBppFromIHDRData(const void* buffer)
+{
+    static const BYTE channels[] = { 1, 0, 3, 1, 2, 0, 4 };
+    const BYTE* p = (BYTE*)buffer, depth = p[8], type = p[8 + 1];
+    return (depth <= 16 && type <= 6) ? channels[type] * depth : 0;
+}
+
+static bool GetInfoFromPng(const void* file, SIZE_T size, IMAGEINFO& info)
+{
+    if (size > 8 + (8 + (4 + 4 + 5) + 4) && IsPngSignature(file))
+    {
+        const UINT PNGIHDRSIG = 0x52444849; // Note: Big endian
+        const UINT* chunkhdr = (UINT*)((char*)file + 8);
+        if (BigToHost32(chunkhdr[0]) >= (4 + 4 + 5) && chunkhdr[1] == PNGIHDRSIG)
+        {
+            info.w = BigToHost32(chunkhdr[2]);
+            info.h = BigToHost32(chunkhdr[3]);
+            info.bpp = GetPngBppFromIHDRData(&chunkhdr[2]);
+            return info.bpp != 0;
+        }
+    }
+    return false;
+}
+
+static bool GetInfoFromBmp(const void* pBitmapInfo, IMAGEINFO& info)
+{
+    BitmapInfoHeader bih(pBitmapInfo);
+    info.w = bih.biWidth;
+    info.h = abs((int)bih.biHeight);
+    UINT bpp = bih.biBitCount * bih.biPlanes;
+    info.bpp = LOBYTE(bpp);
+    return info.w && bpp == info.bpp;
+}
+
+static bool GetInfoFromIcoBmp(const void* pBitmapInfo, IMAGEINFO& stat)
+{
+    bool ret = GetInfoFromBmp(pBitmapInfo, stat);
+    stat.h /= 2; // Don't include mask
+    return ret && stat.h;
+}
+
+EXTERN_C LPCWSTR GetExtraExtensionsGdipList()
```

Comment:
```suggestion
EXTERN_C LPCWSTR GetExtraExtensionsGdipList(VOID)
```
and PCWSTR if possible.